{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "scientific-chuck",
   "metadata": {},
   "source": [
    "# Ets in putative heart enhancers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "appreciated-content",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "proud-stability",
   "metadata": {},
   "outputs": [],
   "source": [
    "def loadEts():\n",
    "    ref  =f'Figure4-parsed_Ets1_8mers.txt'              \n",
    "    Seq2EtsAff  = {line.split('\\t')[0]:float(line.split('\\t')[1]) for line in open(ref,'r').readlines()}\n",
    "    return Seq2EtsAff\n",
    "\n",
    "Ets2Aff=loadEts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "featured-multimedia",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load hg38  genome\n",
    "d=\"/projects/ps-farleylab/users/solvason/ref/genome/\"\n",
    "def loadHg38():\n",
    "    file_genome=f'{d}/human/hg38/hg38.fa'\n",
    "    Handle = open(file_genome, \"r\")\n",
    "    chr2seq = SeqIO.to_dict(SeqIO.parse(Handle, \"fasta\"))\n",
    "    chr2seq = {chrom:str(chr2seq[chrom].seq).upper() for chrom in chr2seq}\n",
    "    return chr2seq\n",
    "\n",
    "# Load 2019 ciona genome\n",
    "def loadCi19():\n",
    "    file_genome=f'{d}/ciona/2019/HT.Ref.fasta'\n",
    "    Handle = open(file_genome, \"r\")\n",
    "    chr2seq = SeqIO.to_dict(SeqIO.parse(Handle, \"fasta\"))\n",
    "    chr2seq = {chrom:str(chr2seq[chrom].seq).upper() for chrom in chr2seq}\n",
    "    return chr2seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "enormous-polls",
   "metadata": {},
   "source": [
    "# Load genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "isolated-windows",
   "metadata": {},
   "outputs": [],
   "source": [
    "G2Chr2Seq={}\n",
    "\n",
    "G2Chr2Seq['Ci19']=loadCi19()\n",
    "G2Chr2Seq['hg38']=loadHg38()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "wooden-sentence",
   "metadata": {},
   "source": [
    "# Analyze ets affinities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "liable-mandate",
   "metadata": {},
   "outputs": [],
   "source": [
    "Ets2RoundedAff={kmer:round(float(aff),2) for kmer,aff in Ets2Aff.items()}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "northern-smith",
   "metadata": {},
   "outputs": [],
   "source": [
    "Genome2Organism={\n",
    "    'Ci19':'Ciona',\n",
    "    'hg38':'Human'\n",
    "}\n",
    "\n",
    "Genome2Name2Bed={\n",
    "    'Ci19':{'Heart-Open-Chromatin-Racioppi'   :f'Figure4-LacZ_6hpf_13_peaks_noexon_top90.bed'}, # modified this to be correct address\n",
    "    \n",
    "    'hg38': {\n",
    "        'Heart-Put-Dev-Enhancer-Multiple-Datasets-No-Dantonio'    :'Figure4-union-putative-dev-cardiac-enhancers_noDantonio.noexons.sorted.bed'\n",
    "    }\n",
    "}\n",
    "\n",
    "xAdjustCustom=[0,+1,-1,-1,+1]# see cell below to adjust x axis so the dots dont overlap on scatter\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d1c8eaa-ebaf-4908-ac7a-e3cf2a3682ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bp_ref_alt(ref8mer,alt8mer):\n",
    "    if len(ref8mer)!=len(alt8mer):\n",
    "        raise ValueError('get_bp_ref_alt(ref8mer,alt8mer)')\n",
    "        \n",
    "    for i,(ref,alt) in enumerate(zip(ref8mer,alt8mer)):\n",
    "        if ref==alt: continue\n",
    "        else:\n",
    "            return i,ref,alt\n",
    "        \n",
    "get_bp_ref_alt('AAGGAATT','AAGGAAAT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3122364a-6d64-4440-86a3-59f192210568",
   "metadata": {},
   "outputs": [],
   "source": [
    "def GenerateAllPossibleSequences(template):\n",
    "        '''Takes a string with letters A,T,C,G,N and returns all possible DNA strings converting N=>A,T,G,C'''\n",
    "        randomSeqList=list(itertools.product('ATGC',repeat=template.count('N')))\n",
    "        seqList=set()\n",
    "        for seqPossibility in randomSeqList:\n",
    "                seq=seqPossibility \n",
    "                dnaOut=''\n",
    "                for i,nt in enumerate(template):\n",
    "                        if nt == 'N':\n",
    "                                dnaOut+=seq[0]\n",
    "                                seq=seq[1:]\n",
    "                        else:\n",
    "                                dnaOut+=nt\n",
    "                seqList.add(dnaOut)\n",
    "        return seqList\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "handled-radical",
   "metadata": {},
   "outputs": [],
   "source": [
    "etsTfCores=set(['GGAA','GGAT','TTCC','ATCC'])\n",
    "\n",
    "\n",
    "def checkIfSnvCausesDramaticIncrease(ref8mer,Ets2RoundedAff,foldChangeCutoff):\n",
    "    \n",
    "    kmer=ref8mer\n",
    "    aff=Ets2RoundedAff[kmer]\n",
    "    \n",
    "    SnvCausesDramaticIncrease=False\n",
    "    DramaticIncreaseFoldChanges=[]\n",
    "    DramaticAffinitySnvs=[]\n",
    "    # If low aff ets, see which snvs cause dramatic affinity increase\n",
    "    for i in range(len(kmer)):\n",
    "\n",
    "        # Create all possible snvs\n",
    "        kmerDegenerate=kmer[:i]+'N'+kmer[i+1:]\n",
    "        core=kmerDegenerate[2:6]\n",
    "        \n",
    "        if core not in etsTfCores: continue \n",
    "            \n",
    "        allSnvs=GenerateAllPossibleSequences(kmerDegenerate)\n",
    "        for snv in allSnvs:\n",
    "\n",
    "            snvAff=Ets2RoundedAff[snv]\n",
    "            foldChange=snvAff/aff\n",
    "\n",
    "            # If affinity jump is sufficient, save \n",
    "            if foldChange>=foldChangeCutoff:\n",
    "                SnvCausesDramaticIncrease=True\n",
    "                DramaticIncreaseFoldChanges.append(foldChange)\n",
    "                DramaticAffinitySnvs.append(get_bp_ref_alt(ref8mer,snv))\n",
    "    \n",
    "    if DramaticIncreaseFoldChanges!=[]:\n",
    "        return SnvCausesDramaticIncrease,max(DramaticIncreaseFoldChanges),DramaticAffinitySnvs\n",
    "    else:\n",
    "        return SnvCausesDramaticIncrease,False,DramaticAffinitySnvs\n",
    "    \n",
    "checkIfSnvCausesDramaticIncrease('AAGGAATT',Ets2RoundedAff,3)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "rural-stuart",
   "metadata": {},
   "source": [
    "# Option 2: Violin + pie chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66e7cf00-4073-4b8e-97c6-b874d337788a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_row(rowList,delim='\\t'):\n",
    "    '''Write a single row of a tsv file.'''\n",
    "    return delim.join([str(i) for i in rowList])+'\\n'\n",
    "\n",
    "def read_tsv(fn,pc,header,breakBool=False,sep='\\t',pc_list=False):\n",
    "    '''Read a tsv file'''\n",
    "    with open(fn,'r') as f:\n",
    "\n",
    "        # If printing columns, skip header\n",
    "        if header:\n",
    "            if (pc==True): pass\n",
    "            else:          next(f)\n",
    "\n",
    "        for i,line in enumerate(f):\n",
    "            a=line.strip().split(sep)\n",
    "            if pc:\n",
    "                if pc_list==False:\n",
    "                    if i==0:\n",
    "                        for i,c in enumerate(a):\n",
    "                            print(i,c)\n",
    "                        print()\n",
    "                        if breakBool: break\n",
    "                        continue\n",
    "                else:\n",
    "                    if i==0:\n",
    "                        print(', '.join([i.replace('-','_').replace(' ','_') for i in a]))\n",
    "                        if breakBool: break\n",
    "                        continue\n",
    "            yield a\n",
    "            \n",
    "def percent(number,rounding_digit=1):\n",
    "    '''Get percent of fraction'''\n",
    "    if rounding_digit==0:\n",
    "        return str(int(100*number))+'%'\n",
    "    else:\n",
    "        return str(round(100*number,rounding_digit))+'%'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3778019-1cb1-4cfd-b713-65079dfd4cfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "def revcomp(dna):\n",
    "        '''Takes DNA sequence as input and returns reverse complement'''\n",
    "        inv={'A':'T','T':'A','G':'C','C':'G', 'N':'N','W':'W'}\n",
    "        revcomp_dna=[]\n",
    "        for nt in dna:\n",
    "                revcomp_dna.append(inv[nt])\n",
    "        return ''.join(revcomp_dna[::-1])\n",
    "    \n",
    "def get_kmers(string,k):\n",
    "        '''Takes DNA sequence as input and a kmer length and YIELDS all kmers of length K in the sequence.'''\n",
    "        for i in range(len(string)):\n",
    "                kmer8=string[i:i+k]\n",
    "                if len(kmer8)==k:\n",
    "                        yield kmer8\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "continued-bulgaria",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def my_autopct(pct):\n",
    "    return (f'{int(pct)}%') if pct > 5 else ''\n",
    "\n",
    "ets=set(['GGAA',revcomp('GGAA'),'GGAT',revcomp('GGAT')])\n",
    "foldChangeCutoff=3\n",
    "\n",
    "\n",
    "skip_ciona=False #changed this to False\n",
    "maxBetaLines=1000\n",
    "\n",
    "allData=[]\n",
    "allDataLabels=[]\n",
    "allAverageEtsAffs=[]\n",
    "\n",
    "\n",
    "line_out_opt_ets_tsv=write_row(['chr','hg38-pos0','hg38-pos1','ref','alt'])\n",
    "\n",
    "for genome in ['Ci19','hg38']:\n",
    "    \n",
    "    if genome=='Ci19' and skip_ciona==True:\n",
    "        continue\n",
    "\n",
    "    print('---------------------------'*3)\n",
    "    print(genome)\n",
    "    print()\n",
    "    \n",
    "    # Set up datasets\n",
    "    Name2KmerSeqAff2Count={}\n",
    "    Name2FoldChanges     ={}\n",
    "    \n",
    "    Name2NPutEns={}\n",
    "    Name2NEtsInPutEns={}\n",
    "    \n",
    "    Name2NPutEnsWithDramaticAffIncrease={}\n",
    "    Name2NEsInPutEnsWithDramaticAffIncrease={}\n",
    "    \n",
    "    #####################################################################################\n",
    "    # Analyze all datasets\n",
    "    #####################################################################################\n",
    "    \n",
    "    Name2Bed=Genome2Name2Bed[genome]\n",
    "\n",
    "    for name,bed in Name2Bed.items():\n",
    "        print('\\t',name,bed)\n",
    "        \n",
    "        # Set up this dataset\n",
    "        Name2KmerSeqAff2Count[name]={}\n",
    "        \n",
    "        Name2NPutEns[name]=0\n",
    "        Name2NEtsInPutEns[name]=0\n",
    "        \n",
    "        Name2NPutEnsWithDramaticAffIncrease[name]=0\n",
    "        Name2NEsInPutEnsWithDramaticAffIncrease[name]=0\n",
    "        \n",
    "        # Read all lines\n",
    "        MaxFoldChangeInPutEn=[]\n",
    "        \n",
    "        # Keep track of length distribution of regions\n",
    "        regionLengthDist=[]\n",
    "        \n",
    "        # Begin reading file\n",
    "        for row in read_tsv(bed,pc=False,header=False):\n",
    "                \n",
    "            chrom,start,end=row[:3]\n",
    "            start0=int(start)\n",
    "            end0=int(end)\n",
    "\n",
    "            regionLengthDist.append(end0-start0)\n",
    "            \n",
    "            Name2NPutEns[name]+=1\n",
    "            thisPutEnHasDramaticAffIncrease=False\n",
    "            maxEtsFoldChangedObserved=0\n",
    "            \n",
    "            \n",
    "            for kmer8_index,kmer8 in enumerate(get_kmers(G2Chr2Seq[genome][chrom][start0:end0],8)):\n",
    "                \n",
    "                thisEtsHasDramaticAffIncrease=False\n",
    "\n",
    "                kmer8=kmer8.upper()\n",
    "                if 'N' in kmer8: continue\n",
    "\n",
    "                core=kmer8[2:6]\n",
    "\n",
    "                if core in ets:\n",
    "                    Name2NEtsInPutEns[name]+=1\n",
    "                    \n",
    "                    aff=Ets2RoundedAff[kmer8]\n",
    "\n",
    "                    if (kmer8,aff) not in Name2KmerSeqAff2Count[name]:\n",
    "                        Name2KmerSeqAff2Count[name][(kmer8,aff)]=0\n",
    "                    Name2KmerSeqAff2Count[name][(kmer8,aff)]+=1\n",
    "                    \n",
    "                    SnvCausesDramaticIncrease,maxFoldChange,DramaticAffinitySnvs=checkIfSnvCausesDramaticIncrease(kmer8,Ets2RoundedAff,foldChangeCutoff)\n",
    "                    \n",
    "                    if SnvCausesDramaticIncrease:\n",
    "                        if maxFoldChange>maxEtsFoldChangedObserved: maxEtsFoldChangedObserved=maxFoldChange\n",
    "                        thisEtsHasDramaticAffIncrease  =True\n",
    "                        thisPutEnHasDramaticAffIncrease=True\n",
    "                        \n",
    "                        for rel_8mer_idx,ref,alt in DramaticAffinitySnvs:\n",
    "                            pos0=start0+kmer8_index+rel_8mer_idx\n",
    "                            pos1=pos0+1\n",
    "                            line_out_opt_ets_tsv+=write_row([chrom,pos0,pos1,ref,alt])\n",
    "                            \n",
    "                        \n",
    "                        \n",
    "                if thisEtsHasDramaticAffIncrease:\n",
    "                    Name2NEsInPutEnsWithDramaticAffIncrease[name]+=1\n",
    "                    \n",
    "            if thisPutEnHasDramaticAffIncrease:\n",
    "                Name2NPutEnsWithDramaticAffIncrease[name]+=1\n",
    "                MaxFoldChangeInPutEn.append(maxEtsFoldChangedObserved)\n",
    "            else:\n",
    "                MaxFoldChangeInPutEn.append(0)\n",
    "            \n",
    "\n",
    "        nPutTotal    = Name2NPutEns[name]\n",
    "        nPutIncrease = Name2NPutEnsWithDramaticAffIncrease[name]\n",
    "        \n",
    "        print(name)\n",
    "        print(f'\\tN Put Enhancers Total                             = {nPutTotal:,}')\n",
    "        print(f'\\tN Put Enhancers With Dramatic Aff Increase        = {nPutIncrease:,} ({percent(nPutIncrease/nPutTotal)})')\n",
    "        print()\n",
    "\n",
    "        nEtsInPutTotal   =Name2NEtsInPutEns[name]\n",
    "        nEtsInPutIncrease=Name2NEsInPutEnsWithDramaticAffIncrease[name]\n",
    "        print(f'\\tN Ets in Put Enhancers Total                      = {nEtsInPutTotal:,}')\n",
    "        print(f'\\tN Ets in Put Enhancers With Dramatic Aff Increase = {nEtsInPutIncrease:,} ({percent(nEtsInPutIncrease/nEtsInPutTotal)})')\n",
    "        print()\n",
    "        \n",
    "\n",
    "\n",
    "        ################################################################\n",
    "        # Determine all possible SNVs which create dramatic change\n",
    "        ################################################################\n",
    "\n",
    "        namePlusSnv='+SNV '+name\n",
    "        Name2KmerSeqAff2Count[namePlusSnv]={}\n",
    "        Name2FoldChanges[namePlusSnv]=[]\n",
    "\n",
    "        for (kmer,aff),count in Name2KmerSeqAff2Count[name].items():\n",
    "\n",
    "            # If low aff ets, see which snvs cause dramatic affinity increase\n",
    "            for i in range(len(kmer)):\n",
    "\n",
    "                # Create all possible snvs\n",
    "                kmerDegenerate=kmer[:i]+'N'+kmer[i+1:]\n",
    "\n",
    "                # Only keep snv if snv is in flank\n",
    "                core=kmerDegenerate[2:6]\n",
    "                if 'N' in core: continue\n",
    "\n",
    "                allSnvs=GenerateAllPossibleSequences(kmerDegenerate)\n",
    "                for snv in allSnvs:\n",
    "\n",
    "                    snvAff=Ets2RoundedAff[snv]\n",
    "                    foldChange=snvAff/aff\n",
    "                    Name2FoldChanges[namePlusSnv].append(snvAff/aff)\n",
    "\n",
    "                    # If affinity jump is sufficient, save \n",
    "                    if foldChange>=foldChangeCutoff:\n",
    "\n",
    "                        if (snv,snvAff) not in Name2KmerSeqAff2Count[namePlusSnv]:\n",
    "                            Name2KmerSeqAff2Count[namePlusSnv][(snv,snvAff)]=0\n",
    "                        Name2KmerSeqAff2Count[namePlusSnv][(snv,snvAff)]+=count\n",
    "\n",
    "        ################################################################\n",
    "        # Violiin plot\n",
    "        ################################################################\n",
    "\n",
    "            \n",
    "        # Data\n",
    "        Name2AffArray={}\n",
    "        for dname in Name2KmerSeqAff2Count:\n",
    "            Name2AffArray[dname]=[]\n",
    "\n",
    "            for (kmer,aff),count in Name2KmerSeqAff2Count[dname].items():\n",
    "                Name2AffArray[dname]+=[aff]*count\n",
    "\n",
    "        snvLabel='+SNVs Opt Aff ≥3-fold\\n'\n",
    "        data=[Name2AffArray[name],Name2AffArray[namePlusSnv]]\n",
    "        \n",
    "        # Save data for further analysis\n",
    "        np.savetxt(name + \"_affinities.txt\", np.array(data[0]), fmt=\"%.3f\")\n",
    "        \n",
    "        allData+=data\n",
    "        \n",
    "        avgOpenAff = round(np.median(Name2AffArray[name]),4)\n",
    "        avgSnvAff  = round(np.median(Name2AffArray[namePlusSnv]),4)\n",
    "\n",
    "        print(f'Average Putative Dev Enhancer ETS     = {avgOpenAff}')\n",
    "        print(f'Average Putative Dev Enhancer ETS-SNV = {avgSnvAff}')\n",
    "        \n",
    "        allAverageEtsAffs+=[avgOpenAff,avgSnvAff]\n",
    "        \n",
    "    \n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff885bfd-4b50-40a4-98ef-7cbdd4b88291",
   "metadata": {},
   "source": [
    "# Plot on same x axis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d535eab7-09d0-4fcc-a1a0-d5b8d081333d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Recall you have the following data\n",
    "# allData=[]\n",
    "# allDataLabels=[]\n",
    "# allAverageEtsAffs=[]\n",
    "\n",
    "for text in [True,False]:\n",
    "\n",
    "    fig,ax=plt.subplots(1,figsize=(11,5),dpi=200)\n",
    "\n",
    "    sns.violinplot(data=allData,ax=ax,scale='area',palette=['lightblue','royalblue']*2,cut=0)\n",
    "\n",
    "    for i,avgOpenAff in enumerate([allAverageEtsAffs[0],'NA',allAverageEtsAffs[2],'NA']):\n",
    "        if avgOpenAff=='NA': continue\n",
    "        ax.hlines(y=avgOpenAff, xmin=i+0, xmax=i+.55, linewidth=2, color='darkgrey',alpha=.3,ls='--')\n",
    "        if text: ax.text(y=avgOpenAff,x=i+.55,s=avgOpenAff,va='center',color='darkgrey')\n",
    "\n",
    "    for i,avgSnvAff in enumerate([allAverageEtsAffs[1],'NA',allAverageEtsAffs[3]]):\n",
    "        if avgSnvAff=='NA': continue\n",
    "        ax.hlines(y=avgSnvAff, xmin=i+1, xmax=i+1.15, linewidth=2,  color='grey',alpha=1,ls='--')\n",
    "        if text: ax.text(y=avgSnvAff,x=i+1.15,s=avgSnvAff,va='center',color='grey')\n",
    "\n",
    "    if text: ax.set_xticklabels(allDataLabels)\n",
    "    if text: ax.set_ylabel('ETS Affinity')\n",
    "    \n",
    "    else:    ax.set_xticklabels([])\n",
    "    if not text: ax.set_yticklabels([])\n",
    "    \n",
    "    plt.show()\n",
    "    \n",
    "    fig.savefig(f\"merged-xaxis.text={text}.violin.png\", format=\"png\",dpi=300)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
